File: core/src/storage/sd_salt.py
from micropython import const

import storage.device
from trezor import fatfs
from trezor.sdcard import with_filesystem
from trezor.utils import consteq

if False:
    from typing import TypeVar, Callable

    T = TypeVar("T", bound=Callable)

SD_CARD_HOT_SWAPPABLE = False
SD_SALT_LEN_BYTES = const(32)
SD_SALT_AUTH_TAG_LEN_BYTES = const(16)


class WrongSdCard(Exception):
    pass


def is_enabled() -> bool:
    return storage.device.get_sd_salt_auth_key() is not None


def compute_auth_tag(salt: bytes, auth_key: bytes) -> bytes:
    from trezor.crypto import hmac

    digest = hmac(hmac.SHA256, auth_key, salt).digest()
    return digest[:SD_SALT_AUTH_TAG_LEN_BYTES]


def _get_device_dir() -> str:
    return "/trezor/device_{}".format(storage.device.get_device_id().lower())


def _get_salt_path(new: bool = False) -> str:
    return "{}/salt{}".format(_get_device_dir(), ".new" if new else "")


@with_filesystem
def _load_salt(auth_key: bytes, path: str) -> bytearray | None:
    # Load the salt file if it exists.
    try:
        with fatfs.open(path, "r") as f:
            salt = bytearray(SD_SALT_LEN_BYTES)
            stored_tag = bytearray(SD_SALT_AUTH_TAG_LEN_BYTES)
            f.read(salt)
            f.read(stored_tag)
    except fatfs.FatFSError:
        return None

    # Check the salt's authentication tag.
    computed_tag = compute_auth_tag(salt, auth_key)
    if not consteq(computed_tag, stored_tag):
        return None

    return salt


@with_filesystem
def load_sd_salt() -> bytearray | None:
    salt_auth_key = storage.device.get_sd_salt_auth_key()
    if salt_auth_key is None:
        return None

    salt_path = _get_salt_path()
    new_salt_path = _get_salt_path(new=True)

    salt = _load_salt(salt_auth_key, salt_path)
    if salt is not None:
        return salt

    # Check if there is a new salt.
    salt = _load_salt(salt_auth_key, new_salt_path)
    if salt is None:
        # No valid salt file on this SD card.
        raise WrongSdCard

    # Normal salt file does not exist, but new salt file exists. That means that
    # SD salt regeneration was interrupted earlier. Bring into consistent state.
    # TODO Possibly overwrite salt file with random data.
    try:
        fatfs.unlink(salt_path)
    except fatfs.FatFSError:
        pass

    # fatfs.rename can fail with a write error, which falls through as an FatFSError.
    # This should be handled in calling code, by allowing the user to retry.
    fatfs.rename(new_salt_path, salt_path)
    return salt


@with_filesystem
def set_sd_salt(salt: bytes, salt_tag: bytes, stage: bool = False) -> None:
    salt_path = _get_salt_path(stage)
    fatfs.mkdir("/trezor", True)
    fatfs.mkdir(_get_device_dir(), True)
    with fatfs.open(salt_path, "w") as f:
        f.write(salt)
        f.write(salt_tag)


@with_filesystem
def commit_sd_salt() -> None:
    salt_path = _get_salt_path(new=False)
    new_salt_path = _get_salt_path(new=True)

    try:
        fatfs.unlink(salt_path)
    except fatfs.FatFSError:
        pass
    fatfs.rename(new_salt_path, salt_path)


@with_filesystem
def remove_sd_salt() -> None:
    salt_path = _get_salt_path()
    # TODO Possibly overwrite salt file with random data.
    fatfs.unlink(salt_path)

File: core/embed/rust/src/sdsalt.rs
use crate::{
    storage::get_device_id,
    trezorhal::{
        fatfs::{close, mount, open, read, rename, unlink, unmount, FATFS, FIL},
        hmac::{hmac_sha256, SHA256_DIGEST_LENGTH},
        sdcard::{is_present, power_off, power_on},
        storage::ExternalSalt,
    },
    ui::model_tt::screens_fw::{insert_sd_card, retry_sd_card, retry_wrong_card},
};
use heapless::String;

pub enum LoadSDResult {
    Ok([u8; 32]),
    WrongSD,
    Error,
}

pub fn consteq(buf1: &[u8], buf2: &[u8]) -> bool {
    let mut diff: usize = buf1.len() - buf2.len();

    for i in 0..buf2.len() {
        diff |= (buf1[i] - buf2[i]) as usize;
    }

    diff == 0
}

pub fn load_salt(auth_key: &[u8], path: String<49>) -> Result<Option<[u8; 32]>, ()> {
    let mut fil: FIL = FIL::default();

    if let Ok(()) = open(&mut fil, path.as_str(), 1) {
        let mut salt = [0_u8; 32];
        let mut stored_tag = [0_u8; 16];

        if read(&mut fil, &mut salt).is_err() {
            return Err(());
        };
        if read(&mut fil, &mut stored_tag).is_err() {
            return Err(());
        };

        let mut result = [0_u8; SHA256_DIGEST_LENGTH as _];

        hmac_sha256(auth_key, &salt, &mut result);
        if close(&mut fil).is_err() {
            return Err(());
        };

        if consteq(&result[..16], &stored_tag) {
            return Ok(Some(salt));
        }
    }
    Ok(None)
}

pub fn load_sd_salt(key: &[u8]) -> LoadSDResult {
    let mut salt_path: String<49> = String::new();
    let mut new_salt_path: String<49> = String::new();

    let dev_id = get_device_id();

    unwrap!(salt_path.push_str("/trezor/device_"));
    unwrap!(salt_path.push_str(dev_id.as_str()));
    unwrap!(salt_path.push_str("/salt"));

    unwrap!(new_salt_path.push_str(salt_path.as_str()));
    unwrap!(new_salt_path.push_str(".new"));
    unwrap!(new_salt_path.push_str("\0"));
    unwrap!(salt_path.push_str("\0"));

    let mut ff: FATFS = FATFS::default();

    if mount(&mut ff).is_err() {
        return LoadSDResult::Error;
    };

    let salt = load_salt(key, salt_path.clone());

    if let Ok(salt) = salt {
        if let Some(salt) = salt {
            return LoadSDResult::Ok(salt);
        }
    } else {
        return LoadSDResult::Error;
    }

    let new_salt = load_salt(key, new_salt_path.clone());
    if let Ok(new_salt) = new_salt {
        if new_salt.is_none() {
            return LoadSDResult::WrongSD;
        }
    } else {
        return LoadSDResult::Error;
    }

    // Normal salt file does not exist, but new salt file exists. That means that
    // SD salt regeneration was interrupted earlier. Bring into consistent state.
    // TODO Possibly overwrite salt file with random data.
    if unlink(salt_path.as_str()).is_err() {
        return LoadSDResult::Error;
    };

    // fatfs rename can fail with a write error, which falls through as an
    // FatFSError.  This should be handled in calling code, by allowing the user
    // to retry.
    if rename(new_salt_path.as_str(), salt_path.as_str()).is_err() {
        return LoadSDResult::Error;
    };

    LoadSDResult::Error
}

pub fn load_sd_salt_wrapper(key: &[u8]) -> LoadSDResult {
    power_on();

    let result = load_sd_salt(key);

    let unmount_res = unmount();
    power_off();

    if unmount_res.is_err() {
        LoadSDResult::Error
    } else {
        result
    }
}

pub fn ensure_sdcard() -> bool {
    while !is_present() {
        let retry = insert_sd_card();
        if !retry {
            return false;
        }
    }
    true
}

pub fn sd_card(key: &[u8]) -> Option<ExternalSalt> {
    loop {
        if !ensure_sdcard() {
            return None;
        }

        let result = load_sd_salt_wrapper(key);

        match result {
            LoadSDResult::Ok(salt) => return Some(salt),
            LoadSDResult::WrongSD => {
                let retry = retry_wrong_card();
                if !retry {
                    return None;
                }
            }
            LoadSDResult::Error => {
                let retry = retry_sd_card();
                if !retry {
                    return None;
                }
            }
        }
    }
}
